File: bitwiseAndOfNumbersRange.py
# problem: https://leetcode.com/problems/bitwise-and-of-numbers-range/
# Runtime: 56 ms, faster than 82.71% of Python3 online submissions for Bitwise AND of Numbers Range.
# Memory Usage: 14.2 MB, less than 55.95% of Python3 online submissions for Bitwise AND of Numbers Range.

class Solution:
    def rangeBitwiseAnd(self, left: int, right: int) -> int:
        if left == right:
            return left
        answer = 0
        i = 0
        while right:
            cur_left = left & 1
            cur_right = right & 1
            left >>= 1
            right >>= 1
            if cur_left != cur_right:
                answer = 0
            answer = (cur_left << i) | answer
            i += 1
        return answer

File: bitwiseAndOfNumbersRange.cpp
// problem: https://leetcode.com/problems/bitwise-and-of-numbers-range/
// Runtime: 4 ms, faster than 92.80% of C++ online submissions for Bitwise AND of Numbers Range.
// Memory Usage: 6 MB, less than 30.66% of C++ online submissions for Bitwise AND of Numbers Range.

class Solution
{
public:
	int rangeBitwiseAnd(int left, int right)
	{
		int shift = 0;
		while (left != right)
		{
			left >>= 1;
			right >>= 1;
			++shift;
		}
		return left << shift;
	}
};